src/darwin/Framework/CHIP/CHIPDeviceControllerStartupParams.mm

```diff
@@ -43,28 +48,296 @@ - (instancetype)initWithKeypair:(id<CHIPKeypair>)nocSigner fabricId:(uint64_t)fa
     return self;
 }
 
+- (instancetype)initWithOperationalKeypair:(id<CHIPKeypair>)operationalKeypair
+                    operationalCertificate:(NSData *)operationalCertificate
+                   intermediateCertificate:(nullable NSData *)intermediateCertificate
+                           rootCertificate:(NSData *)rootCertificate
+                                       ipk:(NSData *)ipk
+{
+    if (!(self = [super init])) {
+        return nil;
+    }
+
+    { // Scope for temporaries
+        // ExtractNodeIdFabricIdFromOpCert needs a TLV-encoded opcert, not a DER-encoded one.
+        uint8_t tlvOpCertBuf[Credentials::kMaxCHIPCertLength];
+        MutableByteSpan tlvOpCert(tlvOpCertBuf);
+        CHIP_ERROR err = Credentials::ConvertX509CertToChipCert(AsByteSpan(operationalCertificate), tlvOpCert);
+        if (err != CHIP_NO_ERROR) {
+            CHIP_LOG_ERROR("Unable to convert operational certificate to TLV: %s", ErrorStr(err));
+            return nil;
+        }
+
+        FabricId fabricId;
+        NodeId unused;
+        err = Credentials::ExtractNodeIdFabricIdFromOpCert(tlvOpCert, &unused, &fabricId);
+        if (err != CHIP_NO_ERROR) {
+            CHIP_LOG_ERROR("Unable to extract fabric id from operational certificate: %s", ErrorStr(err));
+            return nil;
+        }
+        if (fabricId == chip::kUndefinedFabricId) {
```

Comment:
Not sure this can happen

Reply:
Ah, nice catch!  Will remove extraneous check.

And I should use `IsValidFabricId` in general when validating fabric ids.